PerimeterXModule/Internals/CredentialsIntelligence/PxLoginData.cs

```diff
@@ -0,0 +1,231 @@
+            {
+                return ExtractCredentialsFromJson(body, userFieldName, passwordFieldName);
+            } else if (contentType.Contains("x-www-form-urlencoded"))
+            {
+                return ReadValueFromUrlEncoded(body, userFieldName, passwordFieldName);
+            } else if (contentType.Contains("form-data"))
+            {
+                return ExtractValueFromMultipart(body, contentType, userFieldName, passwordFieldName);
+            }
+            
+            return null;
+        }
+
+        private ExtractedCredentials ExtractCredentialsFromJson(string body, string userFieldName, string passwordFieldName) {
+
+            dynamic jsonBody = JSON.DeserializeDynamic(body, PxConstants.JSON_OPTIONS);
+
+            string userValue = PxCommonUtils.ExtractValueFromNestedJson(userFieldName, jsonBody);
+            string passValue = PxCommonUtils.ExtractValueFromNestedJson(passwordFieldName, jsonBody);
+
+            return new ExtractedCredentials(userValue, passValue);
+        }
+
+        private ExtractedCredentials ReadValueFromUrlEncoded(string body, string userFieldName, string passwordFieldName)
+        {
+            var parametersQueryString = HttpUtility.ParseQueryString(body);
+            var parametersDictionary = new Dictionary<string, string>();
+            foreach (var key in parametersQueryString.AllKeys)
+            {
+                parametersDictionary.Add(key, parametersQueryString[key]);
+            }
+
+            return ExtractCredentialsFromDictinary(parametersDictionary, userFieldName, passwordFieldName);
+        }
+
+        private ExtractedCredentials ExtractValueFromMultipart(string body, string contentType, string userFieldName, string passwordFieldName)
+        {
+            var formData = new Dictionary<string, string>();
+
+            var boundary = contentType.Split(';')
+                  .SingleOrDefault(x => x.Trim().StartsWith("boundary="))?
+                  .Split('=')[1];
+
+            var parts = body.Split(new[] { "--" + boundary }, StringSplitOptions.RemoveEmptyEntries);
+
+            foreach (var part in parts)
+            {
+                var lines = part.Split(new[] { "\r\n" }, StringSplitOptions.RemoveEmptyEntries);
+
+                if (part.StartsWith("--"))
+                {
+                    continue;
+                }
+
+                var key = string.Empty;
+                var value = new StringBuilder();
+                foreach (var line in lines)
+                {
+                    if (line.StartsWith("Content-Disposition"))
+                    {
+                        key = line.Split(new[] { "name=" }, StringSplitOptions.RemoveEmptyEntries)[1].Trim('\"');
+                    }
+                    else
+                    {
+                        value.Append(line);
+                    }
+                }
+
+                formData.Add(key, value.ToString());
+            }
```

Comment:
I think this should be separated into a utils function to convert multipart bodies to a dictionary. This could be useful later down the line, and isn't strictly related to CI.